### PR TITLE
 read measures in application/soap+xml content-type

### PR DESCRIPTION
**Pull request by AlvaroVega**
Issue:  #759
solution adopted: Parse the XML into a JSON as the value of the measure (then rely un JEXL for further parsing)

For example: 

```
curl -i -X POST 'http://localhost:7897/iot/json/attrs/data?i=disp4&k=izc9cokegoy7kyfgdd6etipt6' -d '<soap:Envelope
	xmlns:soap="http://www.w3.org/2003/05/soap-envelope">
	<soapenv:Header
		xmlns:soapenv="http://www.w3.org/2003/05/soap-envelope"/>
		<soapenv:Body
			xmlns:soapenv="http://www.w3.org/2003/05/soap-envelope">
			<ns21:notificationEventRequest
				xmlns:ns21="http://myurl.com">
				<ns21:Param1>ABC12345</ns21:Param1>
				<ns21:Param2/>
				<ns21:Date>28/09/2023 11:48:15 +0000</ns21:Date>
				<ns21:NestedAttr>
					<ns21:SubAttr>This is a description</ns21:SubAttr>
				</ns21:NestedAttr>
				<ns21:Status>Assigned</ns21:Status>
				<ns21:OriginSystem/>
			</ns21:notificationEventRequest>
		</soapenv:Body>
	</soap:Envelope>' -H 'content-type: application/soap+xml'
```	
	
Produces:

```
rom=n/a | srv=smartcity | subsrv=/ | msg=Options: {
    "url": "http://iot-orion:1026/v2/entities?options=upsert",
    "method": "POST",
    "headers": {
        "fiware-service": "smartcity",
        "fiware-servicepath": "/"
    },
    "json": {
        "id": "thing:disp4",
        "type": "thing",
        "data": {
            "type": "None",
            "value": {
                "Envelope": {
                    "$": {
                        "xmlns:soap": "http://www.w3.org/2003/05/soap-envelope"
                    },
                    "Header": [
                        {
                            "$": {
                                "xmlns:soapenv": "http://www.w3.org/2003/05/soap-envelope"
                            }
                        }
                    ],
                    "Body": [
                        {
                            "$": {
                                "xmlns:soapenv": "http://www.w3.org/2003/05/soap-envelope"
                            },
                            "notificationEventRequest": [
                                {
                                    "$": {
                                        "xmlns:ns21": "http://myurl.com"
                                    },
                                    "Param1": [
                                        "ABC12345"
                                    ],
                                    "Param2": [
                                        ""
                                    ],
                                    "Date": [
                                        "28/09/2023 11:48:15 +0000"
                                    ],
                                    "NestedAttr": [
                                        {
                                            "SubAttr": [
                                                "This is a description"
                                            ]
                                        }
                                    ],
                                    "Status": [
                                        "Assigned"
                                    ],
                                    "OriginSystem": [
                                        ""
                                    ]
                                }
                            ]
                        }
                    ]
                }
            }
        }
    }
} | comp=IoTAgent
```

### CODE REVIEW

File: lib/bindings/HTTPBinding.js [hunk 1]
```diff
@@ -53,6 +54,7 @@ const { promisify } = require('util');
 const json = promisify(bodyParser.json({ strict: false })); // accept anything JSON.parse accepts.
 const text = promisify(bodyParser.text());
 const raw = promisify(bodyParser.raw());
+const xml = promisify(bodyParser.xml());
```
Comment (rg2011) [inline lib/bindings/HTTPBinding.js:57]: Would you consider removing XML namespace prefixes from tags and attributes? They are not guaranteed to be consistent, they might change from one request to the next. Removing prefixes from tags and attributes would make it easier to parse the message with JEXL later.

This is addressed in https://github.com/Leonidas-from-XIV/node-xml2js/issues/87

```suggestion
const xml2js = require('xml2js');
const xmlStripPrefix = xml2js.processors.stripPrefix;
const xml = promisify(bodyParser.xml({
        xmlParseOptions: {
            // XML namespaces might change from one request to the next.
            // It is useful to remove them from the document,
            // to be able to refer to tags later in JEXL transformations.
            // See https://github.com/Leonidas-from-XIV/node-xml2js/issues/87
            tagNameProcessors: [xmlStripPrefix],
            attrNameProcessors: [xmlStripPrefix]
        }
    })
);
```

File: docs/usermanual.md [hunk 2]
```diff
@@ -98,17 +98,11 @@ hello
...
 then the resulting attribute sent to ContextBroker:
 
-{
-   ...
-   "attrHex": {
-     "value": "68656c6c6f"
-     "type": "<the one used at provisiong time for attrHex attribute>"
-   }
-}
+{ ... "attrHex": { "value": "68656c6c6f" "type": "<the one used at provisiong time for attrHex attribute>" } }
```
Comment (fgalan) [inline docs/usermanual.md:101]:
```suggestion
{ ... "attrHex": { "value": "68656c6c6f", "type": "<the one used at provisiong time for attrHex attribute>" } }
```

File: docs/usermanual.md [hunk 3]
```diff
@@ -316,29 +310,24 @@ attribute IDs `h` and `t`, then humidity measures are reported this way:
...
```
Comment (fgalan) [inline docs/usermanual.md:322]:
```suggestion
{ ... "attrHex": { "value": "68656c6c6f", "type": "<the one used at provisiong time for attrHex attribute>" } }
```

File: lib/bindings/HTTPBinding.js [hunk 4]
```diff
@@ -53,6 +54,19 @@ const { promisify } = require('util');
...
 const text = promisify(bodyParser.text());
 const raw = promisify(bodyParser.raw());
+const xml2js = require('xml2js');
+const xmlStripPrefix = xml2js.processors.stripPrefix;
+const xml = promisify(bodyParser.xml({
+        xmlParseOptions: {
+            // XML namespaces might change from one request to the next.
+            // It is useful to remove them from the document,
+            // to be able to refer to tags later in JEXL transformations.
+            // See https://github.com/Leonidas-from-XIV/node-xml2js/issues/87
```
Comment (fgalan) [inline lib/bindings/HTTPBinding.js:64]: Maybe this should be explained in documentation?

File: docs/usermanual.md [hunk 5]
```diff
@@ -86,9 +86,9 @@ following query parameters:
 It is possible to send a single measure to IoT Platform using an HTTP POST request to the
 `/iot/json/attrs/<attributeName>` and the previously explained query parameters.
 
-In this case, sending a single measure, there is possible to send other kinds of payloads like `text/plain` and
-`application/octet-stream`, not just `application/json`. In case of using `application/octet-stream`, data will be 
-treated as binary data, saved in the attribute maped as hex string. I.E:
+In this case, sending a single measure, there is possible to send other kinds of payloads like `text/plain`,
```
Comment (fgalan) [inline docs/usermanual.md:89]: Maybe some example of SOAP measure should be explained in documentation?

File: lib/bindings/HTTPBinding.js (hunk 4 above)
Comment (AlvaroVega) [reply lib/bindings/HTTPBinding.js:64]: Done in https://github.com/telefonicaid/iotagent-json/pull/771/commits/402e20064ee5bb73c62015b95ec47108f6a97570

File: docs/usermanual.md [hunk 6]
```diff
@@ -98,17 +98,93 @@ hello
...
-     "type": "<the one used at provisiong time for attrHex attribute>"
-   }
-}
+{ ... "attrHex": { "value": "68656c6c6f", "type": "<the one used at provisiong time for attrHex attribute>" } }
+
+Note that every group of 2 character (I.E, the first group, `68`) corresponds to a single ASCII character or byte
+received in the payload (in this case, the value `0x68` corresponds to `h` in ASCII). You can use one of the multiple
+tools available online like [this one](https://string-functions.com/string-hex.aspx)
+
+In case of using `application/soap+xml` a measure like:
```
Comment (fgalan) [inline docs/usermanual.md:107]: Maybe this could be included in an specific subsection?

Comment (fgalan) [reply docs/usermanual.md:107]: (Useful if we need to provide support to send the link to the exact subsection)

**Review by fgalan — APPROVED** [review]
LGTM

**Review by mapedraza — APPROVED** [review]
LGTM

**Pull request merged**